src/transformers/trainer.py

```diff
@@ -666,3 +723,19 @@ def _prediction_loop(
             metrics["loss"] = np.mean(eval_losses)
 
         return PredictionOutput(predictions=preds, label_ids=label_ids, metrics=metrics)
+
+    def distributed_concat(self, arr: np.ndarray, num_total_examples: int) -> np.ndarray:
+        assert self.args.local_rank != -1
+
+        def create_copy():
+            return torch.from_numpy(arr).to(self.args.device)
+
+        output_tensors = [create_copy() for _ in range(torch.distributed.get_world_size())]
+        torch.distributed.all_gather(output_tensors, create_copy())
+
+        concat = torch.cat(output_tensors, dim=0)
+        output = concat.detach().cpu().numpy()
+
+        # truncate the dummy elements added by SequentialDistributedSampler
+        output = output[:num_total_examples]
+        return output
```

Comment:
This seems complex but I'm not sure there's another way to get all results on a single node?

Reply:
@julien-c I might be wrong, but from reading the code, I'd guess that that final slice won't do what you want to. 

- in shuffle mode, it might very well be that the duplicated indices are scattered across batches. Therefore, slicing off the last gathered once is not guaranteed to remove those duplicates
- in non-shuffle mode, I don't think (but not sure) that it is guaranteed that the last device has the last batch or, additionally, that 'all_gather' gathers the tensors in order of devices. I am not sure on this point.

Reply:
This seems like it should get the job done (main training loop still needs to be changed). Also, here `allgather` can be replaced with `gather` since we only need aggregation on the 0th rank process.

Reply:
@BramVanroy This sampler implementation does not have a shuffle option.

Reply:
You're right, I went over it a bit too quick. It would be nice if we could still have shuffle, though, as you state in the other comment.